wasm_utils/src/memory/stack.rs

```diff
@@ -0,0 +1,237 @@
+use memory::{
+    allocation::{AllocationError, Length, WasmAllocation},
+    MemoryBits, MemoryInt, MEMORY_INT_MAX,
+};
+use std::convert::TryFrom;
+
+#[derive(Copy, Clone, Default, Debug, PartialEq)]
+// pub in crate for testing
+pub struct Top(pub(in crate::memory) MemoryInt);
+
+impl From<Top> for MemoryInt {
+    fn from(top: Top) -> Self {
+        top.0
+    }
+}
+
+impl From<Top> for usize {
+    fn from(top: Top) -> Self {
+        Self::from(MemoryInt::from(top))
+    }
+}
+
+impl From<Top> for MemoryBits {
+    fn from(top: Top) -> Self {
+        top.0 as MemoryBits
+    }
+}
+
+#[derive(Copy, Clone, Default, Debug, PartialEq)]
+pub struct WasmStack {
+    // pub in crate for testing
+    pub(in crate::memory) top: Top,
+}
+
+impl WasmStack {
+    // represent the max as MemoryBits type to allow gt comparisons
+    pub fn max() -> MemoryBits {
+        MEMORY_INT_MAX
+    }
+
+    // min compares lt so can be a MemoryInt
+    pub fn min() -> MemoryInt {
+        0
+    }
+
+    // A stack can be initialized by giving the last know allocation on this stack
+    pub fn new() -> WasmStack {
+        WasmStack {
+            top: Top(WasmStack::min()),
+        }
+    }
+
+    pub fn next_allocation(&self, length: Length) -> Result<WasmAllocation, AllocationError> {
+        WasmAllocation::new(MemoryInt::from(self.top()).into(), length)
+    }
+
+    pub fn allocate(&mut self, allocation: WasmAllocation) -> Result<Top, AllocationError> {
+        if MemoryInt::from(self.top()) != MemoryInt::from(allocation.offset()) {
+            Err(AllocationError::BadStackAlignment)
+        } else if MemoryBits::from(self.top()) + MemoryBits::from(allocation.length())
+            > WasmStack::max()
+        {
+            Err(AllocationError::OutOfBounds)
+        } else {
+            // @todo i don't know why we return the old top instead of new one?
```

Comment:
seems like the few places that use .allocate() (`WasmStack::write_in_wasm_memory` and `SinglePageManager::allocate`) do the allocation in 3 parts:
- `next_allocation`
- then `allocate`
- then `WasmAllocation::new` with the old top and new length

That could probably be consolidated into one function that just returns a `WasmAllocation` in one step...but not a big deal. I was curious too.

Reply:
separate = easier to test